ios/NotificationServiceExtension/NotificationService.swift

```diff
@@ -0,0 +1,239 @@
+    guard let payload = notificationContent.userInfo["payload"] as? NSDictionary else {
+      throw ExpError.runtimeError("payload missing")
+    }
+    
+    guard let reportID = payload["reportID"] as? Int64 else {
+      throw ExpError.runtimeError("payload.reportID missing")
+    }
+    
+    guard let reportActionID = payload["reportActionID"] as? String else {
+      throw ExpError.runtimeError("payload.reportActionID missing")
+    }
+    
+    guard let onyxData = payload["onyxData"] as? NSArray else {
+      throw ExpError.runtimeError("payload.onyxData missing" + reportActionID)
+    }
+    
+    guard let reportActionOnyxUpdate = onyxData[1] as? NSDictionary else {
+      throw ExpError.runtimeError("payload.onyxData[1] missing" + reportActionID)
+    }
+    
+    guard let reportActionCollection = reportActionOnyxUpdate["value"] as? NSDictionary else {
+      throw ExpError.runtimeError("payload.onyxData[1].value (report action onyx update) missing" + reportActionID)
+    }
+    
+    guard let reportAction = reportActionCollection[reportActionID] as? NSDictionary else {
+      throw ExpError.runtimeError("payload.onyxData[1].value['\(reportActionID)'] (report action) missing" + reportActionID)
+    }
+    
+    guard let avatarURL = reportAction["avatar"] as? String else {
+      throw ExpError.runtimeError("reportAction.avatar missing. reportActionID: " + reportActionID)
+    }
+    
+    guard let accountID = reportAction["actorAccountID"] as? Int else {
+      throw ExpError.runtimeError("reportAction.actorAccountID missing. reportActionID: " + reportActionID)
+    }
+    
+    guard let person = reportAction["person"] as? NSArray else {
+      throw ExpError.runtimeError("reportAction.person missing. reportActionID: " + reportActionID)
+    }
+    
+    guard let personObject = person[0] as? NSDictionary else {
+      throw ExpError.runtimeError("reportAction.person[0] missing. reportActionID: " + reportActionID)
+    }
+    
+    guard let userName = personObject["text"] as? String else {
+      throw ExpError.runtimeError("reportAction.person[0].text missing. reportActionID: " + reportActionID)
+    }
+    
+    return NotificationData(
+      reportID: reportID,
+      reportActionID: reportActionID,
+      avatarURL: avatarURL,
+      accountID: accountID,
+      userName: userName,
+      title: notificationContent.title,
+      messageText: notificationContent.body,
+      roomName: payload["roomName"] as? String
+    )
+  }
+  
+  @available(iOSApplicationExtension 14.0, *)
+  func createMessageIntent(notificationData: NotificationData) -> INSendMessageIntent {
+    // Initialize only the sender for a one-to-one message intent.
+    let handle = INPersonHandle(value: String(notificationData.accountID), type: .unknown)
+    let avatar = fetchINImage(imageURL: notificationData.avatarURL, reportActionID: notificationData.reportActionID)
```

Comment:
just making sure we could have this empty if any error occurs?

Reply:
Yeah if the image fails to download, `avatar` will be `nil` and just a placeholder image would be shown in the notification